projects/feature-activation/0004-phased-testing.md

```diff
@@ -0,0 +1,206 @@
+- Author: Gabriel Levcovitz <<gabriel@hathor.network>>
+
+# Summary
+[summary]: #summary
+
+This RFC describes a phased procedure for testing the Feature Activation project.
+
+# Motivation
+[motivation]: #motivation
+
+After all changes introduced by the previous Feature Activation RFCs are implemented, the mechanism should be tested with some "fake" features in both the testnet and mainnet environments, and only then we'll be able to use it with confidence in the mainnet to change real features like updating the maximum merkle path length, for example.
+
+# Guide-level explanation
+[Guide-level explanation]: #guide-level-explanation
+
+The testing will consist of 3 phases:
+
+1. NOP features in testnet
+2. Maximum merkle path length update in testnet
+3. NOP features in mainnet
+
+Phase 1 will happen independently, and then Phase 2 and Phase 3 can happen concurrently. Note: updating the maximum merkle path length in mainnet is outside the scope of this project and will be described in a separate RFC.
+
+The NOP features used in Phases 1 and 3 are simply defined in the respective network's settings file, but their activation has no real effect in any computation of the full node, that is, no forks are created no matter whether those features are activated or not. Instead, their states will simply be logged by the full node.
+
+After Phase 1 is complete, Phase 2 will introduce a real change that creates a fork in testnet between nodes that recognize its feature activation, and nodes that do not. At the same time, Phase 3 can start. It is analogous to Phase 1, but on mainnet instead of testnet.
+
+# Reference-level explanation
+[Reference-level explanation]: #reference-level-explanation
+
+In this section, technical details are expanded for what was described above.
+
+## Phase 1
+
+In this phase, 3 NOP features are defined for testnet and each one of them undergo their own feature activation process. Then, we validate that their states progress as expected, both by inspecting their values through the Explorer interfaces, and by inspecting logs via AWS CloudWatch queries.
+
+### NOP features configuration
+
+We'll be using 3 NOP features and define them in `testnet.yml`. We'll test different Feature Activation behavior by configuring different settings for each feature:
+
+1. `NOP_FEATURE_1` is expected to be signaled and activated through bit signaling. Therefore, its `signal_support_by_default` will be set to `True`, so testnet miners (the ones used in `tx-mining-service`) will receive block templates with their signal bit enabled. It'll also have a `minimum_activation_height`.
+2. `NOP_FEATURE_2` is expected to NOT be activated through bit signaling. Instead, it'll have `lock_in_on_timeout=True` and will be "forcefully" activated. (TODO: `MUST_SIGNAL` enforcement is not implemented yet. How do we set bits in testnet miners without using `signal_support_by_default`? What about in mainnet?)
```

Comment:
When we implement mandatory signaling for the `MUST_SIGNAL` phase (which I should do before this Phased Testing), how will we set signal bits for our testnet miners? And how will we do it for mainnet, on Phase 3? We'll have to coordinate with ours partners, as using `signal_support_by_default=False` is part of the testing procedure. They'll have to manually set the bits on their side when `MUST_SIGNAL` is reached (or before).

Reply:
I think such coordination is very complicated. We have to find a way to do it ourselves. Isn't it enough to do it on testnet? I mean, mainnet has a hash power that seems too high for us.

Reply:
Makes sense. I agree we do it only on testnet. Changed in https://github.com/HathorNetwork/rfcs/pull/62/commits/2992fd5d8903b25c3a2605803d87b24e14d77603

Reply:
Thanks!